### PR TITLE
les: refactored serverPeerSet

### PR DESCRIPTION
**Pull request by zsfelfoldi**
This PR removes the old serverPeerSet and uses NodeStateMachine to realize the peer set logic of clients, similarly to what has already been done with clientPeerSet. This change makes the code more consistent by removing multiple custom peer maps used by different mechanisms. It also allows further client side feature improvements (mostly related to incentivization).
The NodeStateMachine is also extended with another ForEach iterator that iterates on nodes having a certain field set.

### CODE REVIEW

File: les/fetcher.go [hunk 1]
```diff
@@ -132,13 +136,9 @@ type lightFetcher struct {
...
 	chaindb ethdb.Database
 	reqDist *requestDistributor
-	peerset *serverPeerSet        // The global peerset of light client which shared by all components
-	chain   *light.LightChain     // The local light chain which maintains the canonical header chain.
-	fetcher *fetcher.BlockFetcher // The underlying fetcher which takes care block header retrieval.
-
-	// Peerset maintained by fetcher
-	plock sync.RWMutex
-	peers map[enode.ID]*fetcherPeer
+	ns      *nodestate.NodeStateMachine // The global peerset of light client which shared by all components
```
Comment (ligi) [inline les/fetcher.go:139]:
```suggestion
	ns      *nodestate.NodeStateMachine // The global peerset of light client which is shared by all components
```

File: les/fetcher.go [hunk 2]
```diff
@@ -132,13 +136,9 @@ type lightFetcher struct {
...
+	chain   *light.LightChain           // The local light chain which maintains the canonical header chain.
+	fetcher *fetcher.BlockFetcher       // The underlying fetcher which takes care block header retrieval.
```
Comment (ligi) [inline les/fetcher.go:141]:
```suggestion
	fetcher *fetcher.BlockFetcher       // The underlying fetcher which takes care of block header retrieval.
```

**Review by ligi — CHANGES_REQUESTED** [review]
Minor suggestions. Btw. Really dislike the variable naming. You don't have to be as extreme as Robert C. Martin suggests “You should name a variable using the same care with which you name a first-born child.” - but one/two letter variable names are really bad IMHO if it's not just a counter.

**Comment by rjl493456442** [comment]
@zsfelfoldi I think it's a bit hard to maintain the code if we just use the nodeStateMachine everywhere. 
I think we can define a few wrappers upon the underlying state machine, e.g. `peerSet`, it would be better.

**Comment by zsfelfoldi** [comment]
@rjl493456442 I'm not sure that adding wrappers around NSM would make it nicer. Which part would you like to wrap?
I agree that it's best to not overuse NSM, I'm just thinking about it as a smarter version of the peerSet. What I did in this PR is mostly just that I replaced the old register/unregister subscription system because it's unnecessary to use two different peer subscription systems.

**Comment by rjl493456442** [comment]
@zsfelfoldi Currently, we just use the NSM directly(e.g for peer handling), because there are some details of using the NSM

e.g.

```go

        // Reject the duplicated peer, otherwise register it to peerset.
	var registered bool
	if err := h.server.ns.Operation(func() {
		if h.server.ns.GetField(p.Node(), clientPeerField) != nil {
			registered = true
		} else {
			h.server.ns.SetFieldSub(p.Node(), clientPeerField, p)
		}
	}); err != nil {
		return err
	}
	if registered {
		return errAlreadyRegistered
	}

```

We can still define the `peerset` for the LES-server based on the NSM, and we can replace these entire code snippets with a single line `peerset.Register(peer)`. It's both easier for reading and maintenance

**Comment by rjl493456442** [comment]
The usage of NSM(e.g `Operation()`, `GetField`, `SetFieldSub` are just too much for the server/handler. We need to remember how to use the NSM correctly all the time.

**Pull request closed without merging**